runners/gradle-plugin/src/main/kotlin/org/jetbrains/dokka/gradle/kotlin/kotlinCompilationUtils.kt

```diff
@@ -6,15 +6,47 @@ import org.jetbrains.kotlin.gradle.dsl.KotlinCommonOptions
 import org.jetbrains.kotlin.gradle.dsl.KotlinMultiplatformExtension
 import org.jetbrains.kotlin.gradle.dsl.KotlinSingleTargetExtension
 import org.jetbrains.kotlin.gradle.plugin.KotlinSourceSet
+import org.jetbrains.kotlin.gradle.plugin.mpp.KotlinCommonCompilation
 
 internal typealias KotlinCompilation =
         org.jetbrains.kotlin.gradle.plugin.KotlinCompilation<KotlinCommonOptions>
 
 internal fun Project.compilationsOf(sourceSet: KotlinSourceSet): List<KotlinCompilation> {
     //KT-45412 Make sure .kotlinSourceSets and .allKotlinSourceSets include the default source set
-    return allCompilationsOf(sourceSet).filter { compilation ->
+    val compilations = allCompilationsOf(sourceSet).filter { compilation ->
         sourceSet in compilation.kotlinSourceSets || sourceSet == compilation.defaultSourceSet
     }
+
+    val hasCompatibilityMetadataVariant = compilations.size >= 2
```

Comment:
> Does it mean that a single target does not have compatibility metadata at all?

I see how the name of the variable might be confusing. I've changed it to `hasAdditionalCommonCompatibilityMetadataVariant` to be more specific, and corrected the comment a little.

The original `filterNot` check was introduced so that `ktor` works. One of its modules has both `main` and `commonMain` compilations, and it fails the build with the error I mentioned in the code comment. Our fix was to filter out the `main` compilation, so only `commonMain` is used and there is no error. However, we did not account for the fact that `main` might be the only compilation available.

Single-target projects in <= 1.8.20 returned both `main` and `commonMain`, so everything worked. But in 1.9.0 it stopped returning `commonMain`, and only returned `main` (no other compilations). But we filtered it out regardless, so we were left with an empty list -> got the error.

So this PR essentially fixes the bug of the previous bugfix - this is what the ktor fix should've been: only filter out `main` if `commonMain` is also present (in other words, if compatibility metadata is enabled), otherwise return whatever compilations are available.

This transitively fixes single-target KMP projects, because they have the `main` compilation now, but there's no guarantee that it will exist in the future. So it's not really a **fix** for single-target KMP projects, more like a happy coincidence that makes it work for the time being.

The check for `compilations.size >= 2` is kind of like a prerequisite / fast path for checking if compatibility metadata is enabled. If the size is 1, from what I understand, it must mean it's disabled.

Reply:
> I've changed it to hasAdditionalCommonCompatibilityMetadataVariant

I think it is clearer.

Also, I mean if we use `main` compilation,  can we have the same problem as with `ktor` for a single target?
Anyway, it is better than supporting no single target at all.

Reply:
> can we have the same problem as with ktor for a single target?

I'm actually not sure, but quite possibly, if one of the dependencies leads to the same problem as in ktor ([stacktrace here](https://gist.github.com/IgnatBeresnev/27c623e531ebc822b033094020900ca1)).

> Anyway, it is better than supporting no single target at all.

Indeed, this is why I don't want to call the PR "Fix single-target multiplatform projects" - it doesn't really fix the problem. This PR introduces a better check for `ktor`, which happens to help with single-target projects, but there are no guarantees